File: codebase/src/cpp/test13/federation/JoinFederationTest.cpp
/*
 *   Copyright 2007 The Portico Project
 *
 *   This file is part of portico.
 *
 *   portico is free software; you can redistribute it and/or modify
 *   it under the terms of the Common Developer and Distribution License (CDDL)
 *   as published by Sun Microsystems. For more information see the LICENSE file.
 *
 *   Use of this software is strictly AT YOUR OWN RISK!!!
 *   If something bad happens you do not have permission to come crying to me.
 *   (that goes for your lawyer as well)
 *
 */
#include "JoinFederationTest.h"

// Register test suite with the global repository
CPPUNIT_TEST_SUITE_REGISTRATION( JoinFederationTest );
CPPUNIT_TEST_SUITE_NAMED_REGISTRATION( JoinFederationTest, "JoinFederationTest" );
CPPUNIT_TEST_SUITE_NAMED_REGISTRATION( JoinFederationTest, "join" );
CPPUNIT_TEST_SUITE_NAMED_REGISTRATION( JoinFederationTest, "federationManagement" );

/////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////// Constructors/Destructors /////////////////////////////////
/////////////////////////////////////////////////////////////////////////////////////////////
JoinFederationTest::JoinFederationTest()
{
}

JoinFederationTest::~JoinFederationTest()
{
}

/////////////////////////////////////////////////////////////////////////////////////////////
/////////////////////////////// Test Setup and Helper Methods ///////////////////////////////
/////////////////////////////////////////////////////////////////////////////////////////////

void JoinFederationTest::setUp()
{
    this->defaultFederate = new Test13Federate( "defaultFederate" );
    this->defaultFederate->quickCreate();
}

void JoinFederationTest::tearDown()
{
	this->defaultFederate->quickResign();
	this->defaultFederate->quickDestroy();
	delete this->defaultFederate;
}

////////////////////////////////////////////////////////////////////////////////////////////
/////////////////////////////////////// Test Methods ///////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////
// TEST: (valid) testJoinFederation() //
////////////////////////////////////////
void JoinFederationTest::testJoinFederation()
{
	try
	{
		// create the FederateAmbassador
		defaultFederate->fedamb = new TestNG6FederateAmbassador( defaultFederate );

		// join the federation
		defaultFederate->rtiamb->joinFederationExecution( defaultFederate->getFederateName(),
		                                                  defaultFederate->SIMPLE_NAME,
		                                                  defaultFederate->fedamb );
	}
	catch( RTI::Exception &e )
	{
		// failure!
		failTest( "Exception joining valid federation: %s", e._reason );
	}
}

////////////////////////////////////////////////
// TEST: testJoinFederationThatDoesNotExist() //
////////////////////////////////////////////////
void JoinFederationTest::testJoinFederationThatDoesNotExist()
{
	try
	{
		// create the FederateAmbassador
		defaultFederate->fedamb = new TestNG6FederateAmbassador( defaultFederate );

		// try and join the federation
		defaultFederate->rtiamb->joinFederationExecution( defaultFederate->getFederateName(),
		                                                  "NoSuchFederation",
		                                                  defaultFederate->fedamb );
		failTestMissingException( "FederationExecutionDoesNotExist",
		                          "joining federation that does not exist" );
	}
	catch( RTI::FederationExecutionDoesNotExist &fedne )
	{
		// success!
	}
	catch( RTI::Exception &e )
	{
		failTestWrongException( "FederationExecutionDoesNotExist", e,
		                        "joining federation that does not exist" );
	}
}

//////////////////////////////////////////////////////
// TEST: testJoinFederationWithNullFederationName() //
//////////////////////////////////////////////////////
void JoinFederationTest::testJoinFederationWithNullFederationName()
{
	try
	{
		// create the FederateAmbassador
		defaultFederate->fedamb = new TestNG6FederateAmbassador( defaultFederate );

		// try and join the federation
		defaultFederate->rtiamb->joinFederationExecution( defaultFederate->getFederateName(),
		                                                  NULL,
		                                                  defaultFederate->fedamb );
		failTestMissingException( "FederationExecutionDoesNotExist",
		                          "joining federation with NULL execution name" );
	}
	catch( RTI::FederationExecutionDoesNotExist &fedne )
	{
		// success!
	}
	catch( RTI::Exception &e )
	{
		failTestWrongException( "FederationExecutionDoesNotExist", e,
		                        "joining federation with NULL execution name" );
	}
}

////////////////////////////////////////////////////
// TEST: testJoinFederationWithNullFederateName() //
////////////////////////////////////////////////////
void JoinFederationTest::testJoinFederationWithNullFederateName()
{
	try
	{
		defaultFederate->rtiamb->joinFederationExecution( NULL,
		                                                  defaultFederate->SIMPLE_NAME,
		                                                  defaultFederate->fedamb );
		failTestMissingException( "RTIinternalError","joining federation with NULL federate name" );
	}
	catch( RTI::RTIinternalError &rtie )
	{
		// success!
	}
	catch( RTI::Exception &e )
	{
		failTestWrongException( "RTIinternalError", e,
		                        "joining federation with NULL federate name" );
	}
}

/////////////////////////////////////////////////////
// TEST: testJoinFederationWithEmptyFederateName() //
/////////////////////////////////////////////////////
void JoinFederationTest::testJoinFederationWithEmptyFederateName()
{
	try
	{
		defaultFederate->rtiamb->joinFederationExecution( "",
		                                                  defaultFederate->SIMPLE_NAME,
		                                                  defaultFederate->fedamb );
		failTestMissingException( "RTIinternalError",
		                          "joining federation with empty federate name" );
	}
	catch( RTI::RTIinternalError &rtie )
	{
		// success!
	}
	catch( RTI::Exception &e )
	{
		failTestWrongException( "RTIinternalError", e,
		                        "joining federation with empty federate name" );
	}
}

//////////////////////////////////////////////////////////
// TEST: testJoinFederationWithNullFederateAmbassador() //
//////////////////////////////////////////////////////////
void JoinFederationTest::testJoinFederationWithNullFederateAmbassador()
{
	try
	{
		defaultFederate->rtiamb->joinFederationExecution( defaultFederate->getFederateName(),
		                                                  defaultFederate->SIMPLE_NAME,
		                                                  NULL );
		failTestMissingException( "RTIinternalError", "joining federation with NULL fedamb" );
	}
	catch( RTI::RTIinternalError &rtie )
	{
		// success!
	}
	catch( RTI::Exception &e )
	{
		failTestWrongException( "RTIinternalError", e, "joining federation with NULL fedamb" );
	}
}

/////////////////////////////////////////////////
// TEST: testJoinFederationWhenAlreadyJoined() //
/////////////////////////////////////////////////
void JoinFederationTest::testJoinFederationWhenAlreadyJoined()
{
	try
	{
		defaultFederate->quickJoin();
		defaultFederate->rtiamb->joinFederationExecution( defaultFederate->getFederateName(),
		                                                  defaultFederate->SIMPLE_NAME,
		                                                  defaultFederate->fedamb );
		failTestMissingException( "FederateAlreadyExecutionMember",
		                          "joining federation when already joined" );
	}
	catch( RTI::RTIinternalError &rtie )
	{
		// success!
	}
	catch( RTI::FederateAlreadyExecutionMember &faem )
	{
		// success!
	}
	catch( RTI::Exception &e )
	{
		failTestWrongException( "FederateAlreadyExecutionMember", e,
		                        "joining federation when already joined" );
	}
}

File: codebase/src/java/test/hlaunit/hla13/federation/JoinFederationTest.java
/*
 *   Copyright 2006 The Portico Project
 *
 *   This file is part of portico.
 *
 *   portico is free software; you can redistribute it and/or modify
 *   it under the terms of the Common Developer and Distribution License (CDDL) 
 *   as published by Sun Microsystems. For more information see the LICENSE file.
 *   
 *   Use of this software is strictly AT YOUR OWN RISK!!!
 *   If something bad happens you do not have permission to come crying to me.
 *   (that goes for your lawyer as well)
 *
 */
package hlaunit.hla13.federation;

import hla.rti.FederateAlreadyExecutionMember;
import hla.rti.FederationExecutionDoesNotExist;
import hla.rti.RTIinternalError;
import hla.rti.RestoreInProgress;
import hla.rti.SaveInProgress;
import hlaunit.hla13.common.Abstract13Test;
import hlaunit.hla13.common.Test13Federate;

import org.testng.annotations.AfterClass;
import org.testng.annotations.AfterMethod;
import org.testng.annotations.BeforeClass;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

@Test(singleThreaded=true, groups={"JoinFederationTest", "basic", "join", "federationManagement"})
public class JoinFederationTest extends Abstract13Test
{
	//----------------------------------------------------------
	//                    STATIC VARIABLES
	//----------------------------------------------------------

	//----------------------------------------------------------
	//                   INSTANCE VARIABLES
	//----------------------------------------------------------

	//----------------------------------------------------------
	//                      CONSTRUCTORS
	//----------------------------------------------------------

	//----------------------------------------------------------
	//                    INSTANCE METHODS
	//----------------------------------------------------------
	@BeforeClass(alwaysRun=true)
	public void beforeClass()
	{
		super.beforeClass();
	}

	@BeforeMethod(alwaysRun=true)
	public void beforeMethod()
	{
		// create a federation that we can test with //
		defaultFederate.quickCreate();
	}
	
	@AfterMethod(alwaysRun=true)
	public void afterMethod()
	{
		defaultFederate.quickResign();
		defaultFederate.quickDestroy();
	}

	@AfterClass(alwaysRun=true)
	public void afterClass()
	{
		super.afterClass();
	}

	///////////////////////////////////////////////////////////////////////////////////////////
	////////////////////////////// Join Federation Test Methods ///////////////////////////////
	///////////////////////////////////////////////////////////////////////////////////////////
	// public int joinFederationExecution( String federateType,
	//                                     String federationExecutionName,
	//                                     FederateAmbassador federateReference )
	//     throws FederateAlreadyExecutionMember,
	//            FederationExecutionDoesNotExist,
	//            SaveInProgress,
	//            RestoreInProgress,
	//            RTIinternalError,
	//            ConcurrentAccessAttempted;

	////////////////////////////////////////
	// TEST: (valid) testJoinFederation() //
	////////////////////////////////////////
	@Test
	public void testJoinFederation()
	{
		try
		{
			// try and join a federation //
			defaultFederate.rtiamb.joinFederationExecution( defaultFederate.federateName,
			                                                defaultFederate.simpleName,
			                                                defaultFederate.fedamb );
		}
		catch( Exception e )
		{
			unexpectedException( "joining federation", e  );
		}
		finally
		{
			// clean up for the next test //
			defaultFederate.quickResign();
		}
	}
	
	
	///////////////////////////////////////////////////////
	// TEST: testJoinFederationWhereFederateNameExists() //
	///////////////////////////////////////////////////////
	@Test
	public void testJoinFederationWhereFederateNameExists()
	{
		defaultFederate.quickJoin();
		
		// try and join a second federate with the same name //
		try
		{
			Test13Federate otherFederate = new Test13Federate( "otherFederate", this );
			otherFederate.rtiamb.joinFederationExecution( defaultFederate.federateName,
			                                              defaultFederate.simpleName,
			                                              otherFederate.fedamb );
			expectedException( FederateAlreadyExecutionMember.class, RTIinternalError.class );
		}
		catch( FederateAlreadyExecutionMember aem )
		{
			// SUCCESS
		}
		catch( RTIinternalError rtie )
		{
			// ALSO VALID
		}
		catch( Exception e )
		{
			wrongException( e, FederateAlreadyExecutionMember.class, RTIinternalError.class );
		}
		finally
		{
			defaultFederate.quickResign();
		}
	}
	
	////////////////////////////////////////////////
	// TEST: testJoinFederationThatDoesNotExist() //
	////////////////////////////////////////////////
	@Test
	public void testJoinFederationThatDoesNotExist()
	{
		// try and join a federation that does no exist //
		try
		{
			defaultFederate.rtiamb.joinFederationExecution( "testFederate",
			                                                "noSuchFederation",
			                                                defaultFederate.fedamb );
			expectedException( FederationExecutionDoesNotExist.class );
		}
		catch( FederationExecutionDoesNotExist dne )
		{
			// SUCCESS
		}
		catch( Exception e )
		{
			wrongException( e, FederationExecutionDoesNotExist.class );
		}
	}
	
	//////////////////////////////////////////////////////
	// TEST: testJoinFederationWithNullFederationName() //
	//////////////////////////////////////////////////////
	@Test
	public void testJoinFederationWithNullFederationName()
	{
		// try and join a federation that does no exist (null name) //
		try
		{
			defaultFederate.rtiamb.joinFederationExecution( "testFederate",
			                                                null, /* null federation name */
			                                                defaultFederate.fedamb );
			expectedException( FederationExecutionDoesNotExist.class );
		}
		catch( FederationExecutionDoesNotExist dne )
		{
			// SUCCESS
		}
		catch( Exception e )
		{
			wrongException( e, FederationExecutionDoesNotExist.class );
		}
	}
	
	////////////////////////////////////////////////////
	// TEST: testJoinFederationWithNullFederateName() //
	////////////////////////////////////////////////////
	@Test
	public void testJoinFederationWithNullFederateName()
	{
		// try and join a federation using null name //
		try
		{
			defaultFederate.rtiamb.joinFederationExecution( null, /* null federate name */
			                                                defaultFederate.simpleName,
			                                                defaultFederate.fedamb );
			expectedException( RTIinternalError.class );
		}
		catch( RTIinternalError rtie )
		{
			// SUCCESS
		}
		catch( Exception e )
		{
			wrongException( e, RTIinternalError.class );
		}
	}
	
	/////////////////////////////////////////////////////
	// TEST: testJoinFederationWithEmptyFederateName() //
	/////////////////////////////////////////////////////
	@Test
	public void testJoinFederationWithEmptyFederateName()
	{
		// try and join a federation using empty name //
		try
		{
			defaultFederate.rtiamb.joinFederationExecution( "", /* empty federate name */
			                                                defaultFederate.simpleName,
			                                                defaultFederate.fedamb );
			expectedException( RTIinternalError.class );
		}
		catch( RTIinternalError rtie )
		{
			// SUCCESS
		}
		catch( Exception e )
		{
			wrongException( e, RTIinternalError.class );
		}
	}
	
	//////////////////////////////////////////////////////////
	// TEST: testJoinFederationWithNullFederateAmbassador() //
	//////////////////////////////////////////////////////////
	@Test
	public void testJoinFederationWithNullFederateAmbassador()
	{
		try
		{
			defaultFederate.rtiamb.joinFederationExecution( "testFederate",
			                                                defaultFederate.simpleName,
			                                                null /* null fedamb */ );
			expectedException( RTIinternalError.class );
		}
		catch( RTIinternalError rtie )
		{
			// SUCCESS
		}
		catch( Exception e )
		{
			wrongException( e, RTIinternalError.class );
		}
	}
	
	/////////////////////////////////////////////////
	// TEST: testJoinFederationWhenAlreadyJoined() //
	/////////////////////////////////////////////////
	@Test
	public void testJoinFederationWhenAlreadyJoined()
	{
		// set up by running a valid join //
		defaultFederate.quickJoin();
		
		// try and join the rtiamb to a second federation //
		try
		{
			defaultFederate.rtiamb.joinFederationExecution( "aValidName",
			                                                defaultFederate.simpleName+"2",
			                                                defaultFederate.fedamb );
			expectedException( FederateAlreadyExecutionMember.class, RTIinternalError.class );
		}
		catch( FederateAlreadyExecutionMember aem )
		{
			// SUCCESS
		}
		catch( RTIinternalError rtie )
		{
			// ALSO VALID
		}
		catch( Exception e )
		{
			wrongException( e, FederateAlreadyExecutionMember.class, RTIinternalError.class );
		}
		finally
		{
			defaultFederate.quickResign();
		}
	}
	
	//////////////////////////////////////////////////
	// TEST: testJoinFederationWhenSaveInProgress() //
	//////////////////////////////////////////////////
	@Test(enabled=false)
	public void testJoinFederationWhenSaveInProgress()
	{
		defaultFederate.quickJoin();
		defaultFederate.quickSaveInProgress( "save" );
		Test13Federate secondFederate = new Test13Federate( "secondFederate", this );
		try
		{
			secondFederate.rtiamb.joinFederationExecution( "secondFederate",
			                                               "JoinFederationTest",
			                                               secondFederate.fedamb );
			expectedException( SaveInProgress.class );
		}
		catch( SaveInProgress sip )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, SaveInProgress.class );
		}
	}

	/////////////////////////////////////////////////////
	// TEST: testJoinFederationWhenRestoreInProgress() //
	/////////////////////////////////////////////////////
	@Test(enabled=false)
	public void testJoinFederationWhenRestoreInProgress()
	{
		defaultFederate.quickJoin();
		defaultFederate.quickRestoreInProgress( "save" );
		Test13Federate secondFederate = new Test13Federate( "secondFederate", this );
		try
		{
			secondFederate.rtiamb.joinFederationExecution( "secondFederate",
			                                               "JoinFederationTest",
			                                               secondFederate.fedamb );
			expectedException( RestoreInProgress.class );
		}
		catch( RestoreInProgress rip )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, RestoreInProgress.class );
		}
	}

	//----------------------------------------------------------
	//                     STATIC METHODS
	//----------------------------------------------------------

}
